### PR TITLE
Rename key concepts

### PR DESCRIPTION
**Pull request by chapati23**
A rename proposal for a few key concepts. 

I also tried unifying the naming as we were using different words for the same things at times.

### CODE REVIEW

File: .solhint.test.json [hunk 1]
```diff
@@ -1,6 +1,5 @@
 {
   "extends": "solhint:recommended",
-  "plugins": ["prettier"],
```
Comment (chapati23) [inline .solhint.test.json:3]: I removed prettier from solhint, cleaner to have it run standalone IMO so solhint can focus on linting, and prettier does the formatting

should also make linting a bit faster

File: src/Oracles.sol [hunk 2]
```diff
@@ -1,40 +1,138 @@
...
+contract Oracles is IOracles, RedstoneConsumerNumericBase {
     using OracleValueLib for OracleValue;
 
-    struct OracleBuffer {
-        uint256[100] medians;
-        OracleBufferInfo bufferInfo;
+    // prettier-ignore
+    struct RateFeed {
+        /// @dev A cyclic buffer of the 100 most recent price reports.
+        /// @dev A price is an unwrapped OracleValue converted to uint256
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev A price is an unwrapped OracleValue converted to uint256.
```

File: src/Oracles.sol [hunk 3]
```diff
@@ -1,40 +1,138 @@
...
 
-    struct OracleBufferInfo {
-        uint8 lastIndex;
-        uint8 windowSize;
-        bool bufferFull;
-        OracleValue windowSum;
-        OracleValue windowAverage;
-        uint40 latestTimestamp;
-        uint8 validityFlags;
+    // prettier-ignore
```
Comment (m-chrzan) [inline src/Oracles.sol:24]: Can we disable only the specific rules we want to override?

File: src/Oracles.sol [hunk 4]
```diff
@@ -1,40 +1,138 @@
...
+    // Designed to fit within 1 storage slot
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    // Designed to fit within 1 storage slot.
```

File: src/Oracles.sol [hunk 5]
```diff
@@ -1,40 +1,138 @@
...
+    struct RateFeedDetails {
+        /**********************/
+        /*  Rate Feed Config  */
+        /**********************/
+
+        /// @dev number of the most recent price values to average over.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev Number of the most recent price values to average over.
```

File: src/Oracles.sol [hunk 6]
```diff
@@ -1,40 +1,138 @@
...
+        uint8 priceWindowSize;
+
+        /// @dev the maximal allowed deviation between reported prices within a batch, expressed as a factor < 1.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev The maximal allowed deviation between reported prices within a batch, expressed as a factor < 1.
```

File: src/Oracles.sol [hunk 7]
```diff
@@ -1,40 +1,138 @@
...
         uint16 allowedDeviation;
+
+        /// @dev the minimal number of data providers that need to have reported a value in the last report
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev The minimal number of data providers that need to have reported a value in the last report.
```

File: src/Oracles.sol [hunk 8]
```diff
@@ -1,40 +1,138 @@
...
         uint8 quorum;
+
+        /// @dev the minimal number of data providers that need to be certain of their value.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev The minimal number of data providers that need to be certain of their value.
```

File: src/Oracles.sol [hunk 9]
```diff
@@ -1,40 +1,138 @@
...
         uint8 certaintyThreshold;
+
+        /// @dev the allowed age of a report before it is considered stale, in seconds.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev The allowed age of a report before it is considered stale, in seconds.
```

File: src/Oracles.sol [hunk 10]
```diff
@@ -1,40 +1,138 @@
...
         uint16 allowedStaleness;
+
+
+        /**********************/
+        /*  Rate Feed Values  */
+        /**********************/
+
+        /// @dev index of the most recently reported price.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev Index of the most recently reported price.
```

File: src/Oracles.sol [hunk 11]
```diff
@@ -1,40 +1,138 @@
...
+        uint8 latestPriceIndex;
+
+        /// @dev true if the buffer of the 100 most recent prices has been filled up at least once.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev True if the buffer of the 100 most recent prices has been filled up at least once.
```

File: src/Oracles.sol [hunk 12]
```diff
@@ -1,40 +1,138 @@
...
+        bool bufferFull;
+
+        /// @dev sum of the last `priceWindowSize` prices.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev Sum of the last `priceWindowSize` prices.
```

File: src/Oracles.sol [hunk 13]
```diff
@@ -1,40 +1,138 @@
...
+        OracleValue priceWindowSum;
+
+        /// @dev average of the last `priceWindowSize` prices (i.e. `priceWindowSum / priceWindowSize`).
+        OracleValue priceWindowAverage;
+
+        /// @dev timestamp of the latest price report.
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev Timestamp of the latest price report.
```

File: src/Oracles.sol [hunk 14]
```diff
@@ -1,40 +1,138 @@
...
+        /// @dev index of the most recently reported price.
+        uint8 latestPriceIndex;
+
+        /// @dev true if the buffer of the 100 most recent prices has been filled up at least once.
+        bool bufferFull;
+
+        /// @dev sum of the last `priceWindowSize` prices.
+        OracleValue priceWindowSum;
+
+        /// @dev average of the last `priceWindowSize` prices (i.e. `priceWindowSum / priceWindowSize`).
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev Average of the last `priceWindowSize` prices (i.e. `priceWindowSum / priceWindowSize`).
```

File: src/Oracles.sol [hunk 15]
```diff
@@ -1,40 +1,138 @@
...
+        OracleValue priceWindowAverage;
+
+        /// @dev timestamp of the latest price report.
+        uint40 latestTimestamp;
+
+        /// @dev a bitmask of the following validity flags:
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// @dev A bitmask of the following validity flags:
```

File: src/Oracles.sol [hunk 16]
```diff
@@ -1,40 +1,138 @@
...
+        /// 0x001 - isFresh => the last price report is fresh
+        /// 0x010 - isCertain => at least `certaintyThreshold` data providers are certain of their price report
+        /// 0x100 - isWithinAllowedDeviation => the latest price is within the allowed deviation
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
        /// 0x100 - isWithinAllowedDeviation => the latest price is within the allowed deviation.
```

File: src/Oracles.sol [hunk 17]
```diff
@@ -1,40 +1,138 @@
...
+        uint8 validityFlags;
     }
 
-    mapping(address => OracleBuffer) public rateFeeds;
+    /// @dev Set of supported rate feed IDs
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    /// @dev Set of supported rate feed IDs.
```

File: src/Oracles.sol [hunk 18]
```diff
@@ -1,40 +1,138 @@
...
+    EnumerableSet.AddressSet private _rateFeedIds;
+
+    /// @dev mapping from rateFeedId to a set of data provider addresses allowed to report prices for `rateFeedId`
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    /// @dev mapping from rateFeedId to a set of data provider addresses allowed to report prices for `rateFeedId`.
```

File: src/Oracles.sol [hunk 19]
```diff
@@ -1,40 +1,138 @@
...
+    mapping(address rateFeedId => EnumerableSet.AddressSet dataProviders)
+        private _rateFeedProviders;
+
+    /// @dev mapping from rateFeedId to a rateFeed containing price reports and metadata
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    /// @dev Mapping from rateFeedId to a rateFeed containing price medians and metadata.
```

File: src/Oracles.sol [hunk 20]
```diff
@@ -1,40 +1,138 @@
...
+        uint256 receivedTimestampMilliseconds,
+        uint256 minAllowedTimestampForNewDataSeconds
+    );
+    error CertaintyThresholdNotReached(
+        uint8 receivedCertainties,
+        uint8 certaintyThreshold
+    );
+    error MinAndMaxValuesDeviateTooMuch(uint256 minVal, uint256 maxVal);
+
+    // TODO: This is a placeholder for the actual implementation base on the 2023 PoC
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    // TODO: This is a placeholder for the actual implementation based on the 2023 PoC.
```

File: src/Oracles.sol [hunk 21]
```diff
@@ -1,40 +1,138 @@
...
+        private _rateFeedProviders;
+
+    /// @dev mapping from rateFeedId to a rateFeed containing price reports and metadata
+    mapping(address rateFeedId => RateFeed rateFeed) private _rateFeeds;
+
+    address internal _currentlyUpdatedRateFeedId;
+
+    /**************************************/
+    /*                                    */
+    /*  RedStone Base Contract Overrides  */
```
Comment (m-chrzan) [inline src/Oracles.sol:87]: Is this comment in the correct location? Are we overriding errors?

File: src/Oracles.sol [hunk 22]
```diff
@@ -1,40 +1,138 @@
...
+        uint256 redstoneValue = getOracleNumericValueFromTxMsg(
+            bytes32(uint256(uint160(rateFeedId)))
+        );
+        rateFeed.details.priceWindowSum = OracleValueLib.fromRedStoneValue(
+            redstoneValue
+        );
 
-    function report(address rateFeedId) external {}
+        // TODO: We still would need to decide how to select the latest data timestamp
+        // Because currently we assume providers provide the same timestamp
```
Comment (m-chrzan) [inline src/Oracles.sol:122]: From what I recall, this is the current behavior of RedStone nodes and that we agreed that we're ok with this mechanism.

File: src/Oracles.sol [hunk 23]
```diff
@@ -72,38 +170,207 @@ contract Oracles is IOracles {
...
+
+        if ((maxVal - minVal) > rateFeed.details.allowedDeviation) {
+            revert MinAndMaxValuesDeviateTooMuch(minVal, maxVal);
+        }
+
+        // In this implementation, we do not require sorted values, but we can add it
+        return NumericArrayLib.pickMedian(values);
+    }
+
+    /// @notice Check each provider is a member of _rateFeedProviders[rateFeedId], revert if not
```
Comment (m-chrzan) [inline src/Oracles.sol]:
```suggestion
    /// @notice Check each provider is a member of _rateFeedProviders[rateFeedId], revert if not.
```

File: src/interfaces/IOracles.sol [hunk 24]
```diff
@@ -1,41 +1,142 @@
...
+ * KEY ENTITIES
+ * @dev Owner         = Mento Governance
+ * @dev Data Provider = Oracle node operator reporting individual price points to RedStone's DDL.
+ * @dev DDL           = RedStone's Data Distribution Layer, an off-chain caching service for price data.
+ * @dev Relayer       = Entities batching together indvidual price points from the DDL and submitting them on-chain.
+ *
+ *
+ * KEY CONCEPTS
+ * @dev Rate Feed     = A specific price feed, e.g. "CELO/USD".
+ * @dev Rate Feed ID  = A unique rate feed identifier calculated as address(uint160(uint256(keccak256("CELOUSD")))).
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
 * @dev Rate Feed ID  = A unique rate feed identifier calculated as e.g. address(uint160(uint256(keccak256("CELOUSD")))).
```

File: src/interfaces/IOracles.sol [hunk 25]
```diff
@@ -1,41 +1,142 @@
...
+ *                          the data provider and indicates their confidence in the reported value. If a more than
+ *                          `certaintyThreshold` providers are certain about a price value, a price report is considered
+ *                          "certain" in our contract. Breakers can be triggered if a report is not certain.
+ * @dev Freshness         = A price report is considered fresh if the latest price value is not older than `allowedStalenessInSeconds`.
+ * @dev Allowed Deviation = The maximum allowed deviation between the lowest and highest price values in a price report.
+ */
 interface IOracles {
     /**
-     * @notice Used to submit a new batch of signed price data.
+     * @notice Main input function through which relayers submit new batched price reports on-chain
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @notice Main input function through which relayers submit new batched price reports on-chain.
```

File: src/interfaces/IOracles.sol [hunk 26]
```diff
@@ -1,41 +1,142 @@
...
+     * @notice Adds a new trusted data provider, i.e. an oracle node operator
+     * @param rateFeedId The rate feed for which the new data provider is allowed to report.
+     * @param provider   The new data provider's address.
+     * @dev Only callable by the owner.
+     */
+    function addDataProvider(address rateFeedId, address provider) external;
+
+    /**
+     * @notice Removes a data provider from being allowed to report for a rate feed.
+     * @param rateFeedId The rate feed from which the data provider should be removed
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @param rateFeedId The rate feed from which the data provider should be removed.
```

File: src/interfaces/IOracles.sol [hunk 27]
```diff
@@ -1,41 +1,142 @@
...
     /**
-     * @notice Sets the allowed deviation for a rate feed.
+     * @notice Sets the allowed price deviation between the lowest and highest price values in a report for a rate feed.
      * @param rateFeedId The rate feed being configured.
-     * @param allowedDeviation The maximal multiplicative deviation allowed
-     * between two values in a report batch, expressed as the numerator of a
-     * fraction over uint16.max.
+     * @param allowedDeviation The difference between the lowest and highest value in a price report.
+     *        Expressed as the numerator of a fraction over uint16.max (65535). I.e., if allowedDeviation was 10_000,
+     *        then the difference between price values can't be greater than 1_000 / 65_535 = 0.015259... ≈ 1.526%
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol:139]: I think the intended mechanism is a bit different: we set a high `allowedDeviation` to restrict the allowed ratio. The value restricts how small the ratio between the lowest and highest price can be. E.g. if it's set to `65_530`, then the lowest value divided by the highest value has to be very close to 1. At least that's how I was thinking about this parameter and its usage.

File: src/interfaces/IOracles.sol [hunk 28]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
-     * @dev Only callable by the owner.
+     * @notice Returns the latest price and validity flags.
+     * @param rateFeedId The rate feed being queried.
+     * @return medianRate The median rate.
+     * @return validityFlags The feed's current validity flags, packed into a uint8.
+     * Specifically:
+     * - Bit 0 (least significant): `isFresh`
+     * - Bit 1: `isCertain`
+     * - Bit 2: `isWithinAllowedDeviation`
+     * - Bits 3-7: unused
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * - Bits 3-7: unused.
```

File: src/interfaces/IOracles.sol [hunk 29]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
-     * @notice Adds a new trusted data provider, i.e. a new offchain oracle node
-     * operator.
-     * @param rateFeedId The rate feed for which the new provider is allowed to
-     * report.
-     * @param provider The new provider's address.
-     * @dev Only callable by the owner.
+     * @notice Returns the current configuration parameters for a rate feed.
+     * @param rateFeedId The rate feed being queried.
+     * @return priceWindowSize           The number of most recent median prices to average over for the final reported price.
+     * @return allowedDeviation          The maximum allowed deviation between the lowest and highest price values in a price report
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @return allowedDeviation          The maximum allowed deviation between the lowest and highest price values in a price report.
```

File: src/interfaces/IOracles.sol [hunk 30]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
-     * denominator being fixed to 1e24.
-     * @param rateFeedId The rate feed whose median rate is queried.
-     * @return numerator The numerator of the median rate.
-     * @return denominator The denominator of the median rate, fixed at 1e24.
-     * @dev The denominator is chosen based on Celo's FixidityLib, which is used
-     * in the legacy SortedOracles oracle. See here:
-     * https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/common/FixidityLib.sol#L26
-     * To get the rate in this contract's internal format, and save a bit of gas
-     * on the consumer side, see `medianRateUint64`.
+     * @notice Passthrough function that calls the new main interface `getExchangeRateFor()`
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @notice Passthrough function that calls the new main interface `getExchangeRateFor()`.
```

File: src/interfaces/IOracles.sol [hunk 31]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
+     * @dev We're ignoring the `lastUpdateTimestamp` as this wasn't part of the old SortedOracles interface
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @dev We're ignoring the `lastUpdateTimestamp` as this wasn't part of the old SortedOracles interface.
```

File: src/interfaces/IOracles.sol [hunk 32]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
+     * @param rateFeedId   The rate feed to fetch the latest price for
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @param rateFeedId   The rate feed to fetch the latest price for.
```

File: src/interfaces/IOracles.sol [hunk 33]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
+     * @return numerator   The numerator of the price
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @return numerator   The numerator of the price.
```

File: src/interfaces/IOracles.sol [hunk 34]
```diff
@@ -67,105 +167,106 @@ interface IOracles {
...
     ) external view returns (uint256 numerator, uint256 denominator);
 
     /**
-     * @notice Returns the median rate as a fixed fraction with 8 decimal digits
-     * after the decimal point.
-     * @param rateFeedId The rate feed whose median rate is queried.
-     * @return medianRate The median rate, expressed as the numerator of a fraction
-     * over 1e8.
+     * @notice Returns the timestamp of the latest price report.
+     * @dev Uses the new interface's `latestTimestamp` cast to uint256
```
Comment (m-chrzan) [inline src/interfaces/IOracles.sol]:
```suggestion
     * @dev Uses the new interface's `latestTimestamp` cast to uint256.
```

File: test/Oracles.t.sol [hunk 35]
```diff
@@ -1,44 +1,48 @@
...
+    address public aRateFeed;
     // solhint-disable-next-line no-empty-blocks
     function setUp() public virtual {
         oracles = new Oracles();
         aRateFeed = address(0x1337);
     }
 }
 
+// solhint-disable-next-line max-line-length
+// Example for a Redstone calldata payload // https://github.com/redstone-finance/redstone-near-connectors/blob/7bc02fc5421200b15da56c33c5c6130130b3ff8a/js/test/integration.test.ts#L17-L37
```
Comment (m-chrzan) [inline test/Oracles.t.sol]:
```suggestion
// Example for a Redstone calldata payload:
// https://github.com/redstone-finance/redstone-near-connectors/blob/7bc02fc5421200b15da56c33c5c6130130b3ff8a/js/test/integration.test.ts#L17-L37
```

**Review by sissnad — APPROVED** [review]

File: src/Oracles.sol (hunk 21 above)
Comment (chapati23) [reply src/Oracles.sol:87]: nope, this was a leftover, had it fixed locally but forgot to push it up 😬

image

File: src/interfaces/IOracles.sol (hunk 27 above)
Comment (chapati23) [reply src/interfaces/IOracles.sol:139]: Aha, then I have misunderstood it, apparently!

It seems to me they'd be functionally equivalent, though, just different way of phrasing it, no?

My understanding of `allowedDeviation` was: _"the highest and lowest value in a price report can not differ by more than `allowedDeviation/65_535` in %"_

Your intent AFAIU was this:

```
@param allowedRatio The allowed difference between the lowest and highest value in a price report.
     *        Expressed as the numerator of a fraction over uint16.max (65_535). I.e., if allowedRatio was 60_000,
     *        then the ratio of the lowest value divided by the highest value in a price report can't be greater than
     *        60_000 / 65_535 = 0.991162137606343
@dev The higher we set `allowedRatio`, the lower the deviation tolerance between the lowest and highest price in a price report is
```

Thinking about this a bit more, to me personally `allowedDeviation` seems more obvious and intuitively auto-completes in my brain to "allowed deviation between prices in a report" whereas `allowedRatio` does not immediately click.

I think we could reconcile this by inverting the concept to mean `allowedDeviation = allowedDeviationNumerator/65_535`

Because intuitively the name suggests that the higher we set it => the higher our deviation tolerance is, and vice versa.

**Summing up, wdyt about this:**

```
@param allowedDeviationNumerator The allowed difference between the lowest and highest value in a price report.
     *        Expressed as the numerator of a fraction over uint16.max (65_535). I.e., if allowedRatio was 1_000,
     *        then the difference between price values can't be greater than 1_000 / 65_535 = 0.015259... ≈ 1.526%
@dev The higher this value is, the higher the deviation tolerance between the lowest and highest price in a price report is
```

Comment (m-chrzan) [reply src/interfaces/IOracles.sol:139]: Sure, that could probably work as well, just thought the onchain math is very simple with the other approach. But yeah, a bit easier to understand the number this way.

Comment (m-chrzan) [reply src/interfaces/IOracles.sol:139]: So to formally define this: the deviation parameter `d` is the numerator of a fraction over `uint16.max` such that the maximal value of a batch (call it `M`) can't be larger than the minimal value (`m`) plus that fraction of `m`. Symbolically: `m * (1 + d/uint16.max) >= M`, or `m + m * (d/uint16.max) >= M`. So the fraction wouldn't be a percentage per se, so we don't need to include an additional division by 100 (as in the explanation in your comment).

Comment (m-chrzan) [reply src/interfaces/IOracles.sol:139]: An alternative would be to use this as a percentage difference parameter, i.e. check for `(M - m) / ((M + m)/2) <= d/uint16.max`.

Comment (chapati23) [reply src/interfaces/IOracles.sol:139]: I think in practice, when configuring new rate feeds, we will likely think & communicate in percent wrt max deviation we want to allow.

I'm wondering if we need the full uint16.max precision here or if giving up some precision for the sake of a simpler interface would make sense here.

I.e. instead of `d/uint16.max` we could make it `d/10_000` which would allow for a more intuitive handling.

I'm imagining the input from nadiem/roman to be "max deviation should be 0.5%" so the `allowedDeviation` param could just be `50` instead of `327` (327/65535 = 0.5%)
